_posts/2018-03-01-Rust-1.24.1.md

```diff
@@ -0,0 +1,257 @@
+> cleared in Lua and jumps back to the `lua_pcall` that originated that
+> execution, passing along the error message.
+
+So, the question becomes: Why does this break? And why does it break on
+Windows?
+
+When we talked about `setjmp`/`longjmp` inititally, a key phrase here wasn't
+highlighted. Here it is:
+
+> After digging in, the culpurit was found: `setjmp`/`longjmp`. These functions
+> are *provided by the C standard library* as a means of handling errors. You
+
+These functions aren't part of the C language, but part of the standard
+library. That means that platform authors implement these functions, and
+their implementations may differ.
+
+Windows has a concept called SEH, short for ["Structured Exception
+Handling"](https://msdn.microsoft.com/en-us/library/windows/desktop/ms680657(v=vs.85).aspx).
+Windows uses SEH to implement `setjmp`/`longjmp`, as the whole idea of SEH
+is to have uniform error handling. For similar reasons, C++ exceptions use
+SEH, as do Rust panics.
+
+Before we can sort the exact details of what's happening, let's look at how `rlua`
+works. `rlua` has an internal function, `protect_lua_call`, used to call into
+Lua. Using it looks like this:
+
+```rust
+protect_lua_call(self.state, 0, 1, |state| {
+    ffi::lua_newtable(state);
+})?;
+```
+
+That is, `protect_lua_call` takes some arguments, one of which is a closure. This
+closure is passed to `lua_pcall`, which catches any `longjmp`s that may be thrown
+by the code passed to it, aka, that closure.
+
+Consider the code above, and imagine that `lua_newtable` here could call
+`longjmp`. Here's what should happen:
+
+1. `protect_lua_call` takes our closure, and passes it to `lua_pcall`.
+2. `lua_pcall` calls `setjmp` to handle any errors, and invokes our closure.
+2. Inside our closure, `lua_newtable` has an error, and calls `longjmp`.
+3. The initial `lua_pcall` catches the `longjmp` with the `setjmp` it called earlier.
+4. Everyone is happy.
+
+However, `lua_newtable` is an `extern fn` to Rust. So, with the chances in
+1.24.1, it sets up a panic handler that will cause an abort. In other words,
+the code sorta looks something like this psuedo code now:
+
+```rust
+protect_lua_call(self.state, 0, 1, |state| {
+    let result = panic::catch_unwind(|| {
+        ffi::lua_newtable(state);
+    });
+
+    if result.is_err() {
+        process::abort();
+    }
+})?;
+```
+
+Earlier, when discussing `setjmp`/`longjmp`, we said that the issue with it in
+Rust code is that it doesn't handle Rust destructors. So, on every platform but
+Windows, the above `catch_unwind` shenanigans is effectively ignored, so
+everything works. However, on Windows, since both `setjmp`/`longjmp` and Rust
+panics use SEH, the `longjmp` gets "caught", and runs the new abort code!
+
+The [solution here](https://github.com/rust-lang/rust/pull/48572) is to
+generate the abort handler, but in a way that `longjmp` won't trigger. Eventually,
```

Comment:
This sentence is missing an "it"; it should end "...`longjmp` won't trigger **it**." Alternatively, "...won't trigger the handler."